internal/job/executor.go

```diff
@@ -1185,6 +1186,57 @@ func (e *Executor) setupRedactors() {
 func (e *Executor) kubernetesSetup(ctx context.Context, k8sAgentSocket *kubernetes.Client) error {
 	e.shell.Commentf("Using Kubernetes support")
 
+	rtr := roko.NewRetrier(
+		roko.WithMaxAttempts(7),
+		roko.WithStrategy(roko.Exponential(2*time.Second, 0)),
+	)
+	regResp, err := roko.DoFunc(ctx, rtr, func(rtr *roko.Retrier) (*kubernetes.RegisterResponse, error) {
+		return k8sAgentSocket.Connect(ctx)
+	})
+	if err != nil {
+		return fmt.Errorf("error connecting to kubernetes runner: %w", err)
+	}
+
+	// Set our environment vars based on the registration response.
+	// But note that the k8s stack interprets the job definition itself,
+	// and sets a variety of env vars (e.g. BUILDKITE_COMMAND) that
+	// *could* be different to the ones the agent normally supplies.
+	// Examples:
+	// * The command container could be passed a specific
+	//   BUILDKITE_COMMAND that is computed from the command+args
+	//   podSpec attributes (in the kubernetes "plugin"), instead of the
+	//   "command" attribute of the step.
+	// * BUILDKITE_PLUGINS is pre-processed by the k8s stack to remove
+	//   the kubernetes "plugin". If we used the agent's default
+	//   BUILDKITE_PLUGINS, we'd be trying to find a kubernetes plugin
+	//   that doesn't exist.
+	// So we should skip setting any vars that are already set, and
+	// specifically any that could be deliberately *unset* by the
+	// k8s stack (BUILDKITE_PLUGINS could be unset if kubernetes is
+	// the only "plugin" in the step).
+	// (Maybe we could move some of the k8s stack processing in here?)
+	//
+	// To think about: how to obtain the env vars early enough to set
+	// them in ExecutorConfig (because of how urfave/cli works, it
+	// must happen before App.Run, which is before the program even knows
+	// which subcommand is running).
```

Comment:
Just for my knowledges, what environment variables were we trying to set "early"? 

For our original goal of BUILDKITE_AGEND_ID, does this setup work?

Reply:
Anything that normally configures the bootstrap subcommand, but especially the required flags/vars: `BUILDKITE_JOB_ID`, `BUILDKITE_REPO`, etc. I would have expected the command container to use these without having to set the env vars for them at schedule time. But that didn't work.

For the original goal of agent env vars that aren't known when the pod is scheduled, yes, this should still work. Commands, plugins etc would be run with the `e.shell.Env`, but for completeness this PR still inserts them into the current process env (`os.Setenv`).